### DISCUSSION TITLE
How to implement confirmable coap packets

### DISCUSSION THREAD

**Discussion by ksathara** [openthread/openthread#10323] [Q&A]
Hi,
I have a set of coap clients (SEDs) and set of coap servers (FTDs). The clients periodically send a data packet to a previously provisioned server.  I want to implement confirmable data packets from the clients to server, so that if the data packet reach the provisioned server, we  get ack and if the data packets doesnt reach the prov server, we get NAK or nothing. This way we can detect whether the prov server is reachable from the client's side. 

I implemented a confirmable packet send to the provisioned server and I can see the packets being delivered. 
I send the the packet as follows.
`error = otCoapSendRequest (srv_context.ot, response, &sensorMsgInfo, sensor_data_ack, NULL)`

sensor_data_ack () is my call back for ack. It is defined as 
`int *sensor_data_ack (const struct coap_packet *response, struct coap_reply *reply, const struct sockaddr *from)`

Issues
1. I can see that the call back is invoked time to time. But when I look into 'from' sock address, I cant see ip addr of the prov server. Also the  sa_family in the 'from' is 0x5c00. I think this should be 2 (AE_INET6)?
2. How can I verify the ack for the confirmable packet via the wireshark?
3. Do we need to implement the send ack handler on the server side to handle confirmable messages? 
Cheers,
Kaushalya

**Comment by jwhui** [comment]
> I can see that the call back is invoked time to time. But when I look into 'from' sock address, I cant see ip addr of the prov server. Also the sa_family in the 'from' is 0x5c00. I think this should be 2 (AE_INET6)?

I believe the callback should have type [`otCoapResponseHandler`](https://openthread.io/reference/group/api-coap#otcoapresponsehandler). Also be sure to check the error value passed to the callback.

> How can I verify the ack for the confirmable packet via the wireshark?

Wireshark includes CoAP dissectors.

> Do we need to implement the send ack handler on the server side to handle confirmable messages?

When using OpenThread's CoAP implementation on the server side, the server side is responsible for generating and sending the CoAP response.

**Reply by ksathara** [reply]
Hi Jonathan,
Thanks again. After fixing the callback, I always get OT_ERROR_RESPONSE_TIMEOUT. So seems the ACK is not received. 
> the server side is responsible for generating and sending the CoAP response

What you mean is that we need to send an ACK packet from our application, when we receive a data packet? I thought this is handled by CoAP server code itself. 

If we have to write our own ACK code, are there any examples or documentation how to do it?

> Wireshark includes CoAP dissectors.

Is this a CoAP ACK packet sent from 0x7400 to 0x3000?
![image](https://github.com/openthread/openthread/assets/20408595/feb27138-dd69-4972-9e1a-3d33c1fe6141)

Cheers,
Kaushalya

**Reply by jwhui** [reply]
The packet you highlighted does look like a CoAP ACK message.

You can see the OT CLI example, where the server is generating the CoAP ACK:
https://github.com/openthread/openthread/blob/fcd3c5c4b75cbb67db902e1a241de5be1a284b27/src/cli/cli_coap.cpp#L959
